### PR TITLE
Add Fisher functionality

### PR DESCRIPTION
**Pull request by williamkreamer**
@patricksnape @jabooth just opening this pull request to let you know I'm starting work on this feature

### CODE REVIEW

**Comment by patricksnape** [comment]
@williamkreamer Sounds great - will kick Jenkins off when Travis starts passing - no need to waste the Mac Mini cycles :smile:

**Comment by williamkreamer** [comment]
I've checked in a bunch of stuff. I doubt anything passes. Any tips on how to investigate the build errors to see what I've done wrong?

File: cyvlfeat/fisher/cyfisher.pyx [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+                np.ndarray[float, ndim=2, mode='c'] MEANS,
+                np.ndarray[float, ndim=2, mode='c'] COVARIANCES,
+                np.ndarray[float, ndim=1, mode='c'] PRIORS,
+                bint Normalized,
+                bint SquareRoot,
+                bint Improved,
+                bint Fast,
+                bint Verbose):
+
+    vl_size numClusters = MEANS.shape[1];
```
Comment (patricksnape) [inline cyvlfeat/fisher/cyfisher.pyx]: When you declare typed variables in Cython, you need to prepend them with `cdef` e.g. `cdef vl_size numClusters`

File: cyvlfeat/fisher/cyfisher.pyx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    vl_size dimension = MEANS.shape[0];
+    vl_size numData = X.shape[1];
+    int flags = 0;
+
+    if Normalized:
+        flags |= VL_FISHER_FLAG_NORMALIZED
```
Comment (patricksnape) [inline cyvlfeat/fisher/cyfisher.pyx]: These flags are presumably declared somewhere in a vlfeat header, and need to be declared as the `vl_fisher_encode` was, in that header file. They then need to be `cimported` here.

File: cyvlfeat/fisher/cyfisher.pyx [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        flags |= VL_FISHER_FLAG_SQUARE_ROOT
+        
+    if Improved:
+        flags |= VL_FISHER_FLAG_IMPROVED
+        
+    if Fast:
+        flags |= VL_FISHER_FLAG_FAST
+    
+    if Verbose:
+        print 'vl_fisher: num data: %d' % numData
```
Comment (patricksnape) [inline cyvlfeat/fisher/cyfisher.pyx]: Use `print` as a function e.g. `print('Message')`

File: cyvlfeat/fisher/cyfisher.pyx [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                np.ndarray[float, ndim=1, mode='c'] PRIORS,
+                bint Normalized,
+                bint SquareRoot,
+                bint Improved,
+                bint Fast,
+                bint Verbose):
+
+    vl_size numClusters = MEANS.shape[1];
+    vl_size dimension = MEANS.shape[0];
+    vl_size numData = X.shape[1];
```
Comment (patricksnape) [inline cyvlfeat/fisher/cyfisher.pyx]: The semi-colons aren't required :)

File: cyvlfeat/fisher/cyfisher.pyx [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    if Verbose:
+        print 'vl_fisher: num data: %d' % numData
+        print 'vl_fisher: num clusters: %d' % numClusters
+        print 'vl_fisher: data dimension: %d' % dimension
+        print 'vl_fisher: code dimension: %d' % 2 * numClusters * dimension
+        print 'vl_fisher: square root: %d' % SquareRoot
+        print 'vl_fisher: normalized: %d' % Normalized
+        print 'vl_fisher: fast: %d' % Fast
+
+    enc = np.zeros((2*numClusters*dimension,),dtype=np.float32)
```
Comment (patricksnape) [inline cyvlfeat/fisher/cyfisher.pyx]: You can type this variable e.g. `cdef float[:] enc = np.zeros((2*numClusters*dimension,),dtype=np.float32)`

File: cyvlfeat/fisher/cyfisher.pyx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    numTerms = vl_fisher_encode(&enc.types,
```
Comment (patricksnape) [inline cyvlfeat/fisher/cyfisher.pyx]: You can type `numTerms` as well

File: cyvlfeat/_vl/fisher.pxd [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+from .host cimport vl_type
+
+
+cdef extern from "vl/fisher.h":
+    vl_size vl_fisher_encode(void * enc,
+                             vl_type dataType,
+                             void const * means,
+                             vl_size dimension,
+                             vl_size numClusters,
+                             void const * covariances,
```
Comment (patricksnape) [inline cyvlfeat/_vl/fisher.pxd]: I haven't written any Cython in a while, but I think that Cython gets unhappy about the `const` on pointer types, so try just removing them if this is causing issues.

**Comment by patricksnape** [comment]
Your latest changes looks great, sorry for not warning you about the setup.py, but what you have now looks correct. Now travis-ci.org is giving some hints about what is going wrong, mostly the const thing I think. It's nearly midnight in the UK but tomorrow morning I'll check back and see if you are still having build errors!

**Comment by williamkreamer** [comment]
Thanks, I got the code to build but I think I'm kind of stuck now. It crashes in my unit test with a "bus error" yikes.

**Comment by williamkreamer** [comment]
Hi @patricksnape I've got the code building and I fixed some obvious ctypes errors. I think it should be working now, but I'm still getting a "bus error"

Do you have any suggestions for how I can debug this?

(env)Williams-MacBook-Pro:cyvlfeat williamkreamer$ DYLD_LIBRARY_PATH=/Users/williamkreamer/anaconda/pkgs/vlfeat-0.9.20-0/lib nosetests -v .
cyvlfeat.fisher.tests.fisher_test.test_fisher ... Bus error: 10
(env)Williams-MacBook-Pro:cyvlfeat williamkreamer$

**Comment by williamkreamer** [comment]
Hmmm - when I run just the new fisher test case I see "segmentation error". This smells like I made a mistake in Cython and am seeing a segfault in that code. 

(env)Williams-MacBook-Pro:cyvlfeat williamkreamer$ DYLD_LIBRARY_PATH=/Users/williamkreamer/anaconda/pkgs/vlfeat-0.9.20-0/lib nosetests -v cyvlfeat/fisher/tests/fisher_test.py
cyvlfeat.fisher.tests.fisher_test.test_fisher ... Segmentation fault: 11
(env)Williams-MacBook-Pro:cyvlfeat williamkreamer$

**Comment by williamkreamer** [comment]
Hmmm - tried commenting out the call to vl_fisher_encode (in cyfisher.pyx) and the segfault went away...

Maybe I'm making a mistake in getting data out of the numpy arrays into C

**Comment by patricksnape** [comment]
Unfortunately we've worked over top of each other - but check the PR I made on your fork that should fix these issues. If you merge those in, then they should appear here. There were a few PEP 8 fixes/Python style things, but I also fixed the buffer interfaces so that the sef fault stops. Try fixing the merge conflicts and see if it works for you!

**Comment by patricksnape** [comment]
@williamkreamer If you want to chat more interactively - we can talk [here](https://gitter.im/menpo/cyvlfeat?utm_source=share-link&utm_medium=link&utm_campaign=share-link) if you like!

**Comment by patricksnape** [comment]
Jenkins, test this please

**Comment by williamkreamer** [comment]
I added a test to the fisher unit test suite. Unfortunately, it isn't passing. See the attached image
encoding

**Comment by patricksnape** [comment]
So, this is where it gets tricky. You have to be very careful transferring data between Python and Matlab because of the Fortran/C-ordering of Matlab vs Python. Try building your data matrix in Python using explicit array notation rather than reshaping to try and ensure the input and output is identical. Also, maybe try generating that expected output using all flags as ``False``? Then we can check the flags are getting propagated correctly as well.

**Comment by williamkreamer** [comment]
@patricksnape All set now. I added a second test case that I generated by running vl_fisher_encode from a C program. The python wrapper produces the same result. I think it is ready for your review.

**Comment by jabooth** [comment]
thanks @williamkreamer, just a heads up that @patricksnape is out for the evening now (6:30pm here in London) but I'm sure he'll review this in the morning for you. 

Thanks again for the contribution! :+1:

**Comment by patricksnape** [comment]
@williamkreamer Thanks a lot - this all looks great. There are a couple of very minor stylistic things that I'm just going to change - but to try and make sure that you have the builds when you start work today I'm going to merge this!

**Comment by williamkreamer** [comment]
awesome, thanks!

**Pull request merged**